### PR TITLE
Solution: Intro to Python I

### PR DESCRIPTION
**Pull request by nickcannariato**
## Goals

* Learn the basic syntax and structure of Python

## Summary

* Implement a number of tiny Python programs that demonstrate Python syntax and
  language concepts.

## Instructions

Each directory inside the `src/` directory presents exercises revolving around a
particular concept in Python. Not all of these concepts are unique to Python (in
fact, most probably aren't). This means that you can leverage knowledge you've
obtained via exposure to other programming languages towards learning Python. 

The suggested order for going through each of the directories is: 

* `hello` -- Hello world
* `bignum` -- Print some big numbers
* `datatypes` -- Experiment with type conversion
* `modules` -- Learn to import from modules
* `printing` -- Formatted print output
* `lists` -- Python's version of arrays
* `tuples` -- Immutable lists typically for heterogenous data
* `slices` -- Accessing parts of lists
* `comprehensions` -- List comprehensions
* `dictionaries` -- Dictionaries
* `functions` -- Functions
* `args` -- Arguments and Keyword Arguments
* `scopes` -- Global, Local, and Non-Local scope
* `file_io` -- Read and write from files
* `cal` -- Experiment with module imports and implement a text-based calendar
* `classes` -- Classes and objects

## Stretch Goals

1. One of Python's main philosophical tenets is its emphasis on readability. To
   that end, the Python community has standardized around a style guide called
   [PEP 8](https://www.python.org/dev/peps/pep-0008/). Take a look at it and
   then go over the code you've written and make sure it adheres to what PEP 8
   recommends. Alternatively, PEP 8 linters exist for most code editors (you can
   find instructions on installing a Python linter for VSCode
   [here](https://code.visualstudio.com/docs/python/linting)). Try installing
   one for your editor!

2. Rewrite code challenges you've solved before or projects you've implemented
   before in a different language in Python. Start getting in as much practice
   with the language as possible!

3. Write a program to determine if a number, given on the command line, is prime.

   1. How can you optimize this program?
   2. Implement [The Sieve of
      Erathosthenes](https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes), one
      of the oldest algorithms known (ca. 200 BC).

### CODE REVIEW

File: src/11_args.py [hunk 1]
```diff
@@ -1,17 +1,22 @@
...
+  return x + y
 
 print(f1(1, 2))
 
 # Write a function f2 that takes any number of integer arguments and prints the
 # sum. Google for "python arbitrary arguments" and look for "*args"
 
 # YOUR CODE HERE
+def f2(*args):
+  return reduce(lambda x,y: x+y, args)
```
Comment (jamespagedev) [inline src/11_args.py]: nice job testing new library. I would have just used `return sum(argv)`, but it's good to try out different libraries and get some deep coding practice.

File: src/13_file_io.py [hunk 2]
```diff
@@ -4,15 +4,48 @@
...
 # YOUR CODE HERE
 
+
+def read_and_print(file_path):
+    with open(file_path, 'r') as f:
+        for line in f:
+            print(line)
+
+
+read_and_print('src/foo.txt')
```
Comment (jamespagedev) [inline src/13_file_io.py]: I would avoid hard coding paths with files in the future. you can use the following snippet to grab the path...
```
import os
dirname = os.path.dirname(os.path.abspath(__file__))
filenameFoo = 'foo.txt'
filenameFooWithPath = os.path.join(dirname, filenameFoo)
```

File: src/09_dictionaries.py [hunk 3]
```diff
@@ -35,10 +35,19 @@
...
+    dict(lat=42, lon=-150, name="home")
+)
 
 # Modify the dictionary with name "a place" such that its longitude
 # value is -130 and change its name to "not a real place"
 # YOUR CODE HERE
+for location in waypoints:
+    if location['name'] == 'a place':
+        location['lon'] = -130
+        location['name'] = 'not a real place'
```
Comment (jamespagedev) [inline src/09_dictionaries.py]: Good opportunity here to check out `.update()` submethod for python lists.

File: src/04_printing.py [hunk 4]
```diff
@@ -11,7 +11,10 @@
 # Using the printf operator (%), print the following feeding in the values of x,
 # y, and z:
 # x is 10, y is 2.25, z is "I like turtles!"
+print('x is %d, y is %f, z is "%s"' % (x, y, z))
```
[the remark this answers is not in the document]
Comment (nickcannariato) [reply src/04_printing.py]: Ah, solid feedback. I wasn't aware that you could limit the number of decimal places this way. Corrected in 85e986b

File: src/04_printing.py [hunk 5]
```diff
@@ -11,7 +11,10 @@
...
 
 # Use the 'format' string method to print the same thing
+print('x is {}, y is {}, z is "{}"'.format(x, y, z))
```
[the remark this answers is not in the document]
Comment (nickcannariato) [reply src/04_printing.py]: Corrected in 85e986b

File: src/04_printing.py [hunk 6]
```diff
@@ -11,7 +11,10 @@
...
 
-# Finally, print the same thing using an f-string
\ No newline at end of file
+# Finally, print the same thing using an f-string
+print(f'x is {x}, y is {y}, z is "{z}"')
```
[the remark this answers is not in the document]
Comment (nickcannariato) [reply src/04_printing.py]: Corrected in 85e986b

File: src/11_args.py (hunk 1 above)
Comment (nickcannariato) [reply src/11_args.py]: Thanks! Yeah, `sum` is likely the better choice here, it just seemed like a good place to play with functools a bit and get comfortable with some more disparate tooling in the standard library.

File: src/09_dictionaries.py (hunk 3 above)
Comment (nickcannariato) [reply src/09_dictionaries.py]: Ah, thank you for that. The multi-assignment statement didn't feel right but I wasn't sure what the most appropriate method was. `.update()` makes a ton of sense. 

Fixed in c36a560

**Pull request merged**